src/leader_schedule.rs

```diff
@@ -0,0 +1,46 @@
+use crate::leader_scheduler::LeaderScheduler;
+
+use solana_sdk::pubkey::Pubkey;
+
+pub struct LeaderSchedule {
+    slot_leaders: Vec<Pubkey>,
+    ticks_per_slot: u64,
+    ticks_per_epoch: u64,
+    current_epoch: u64,
```

Comment:
I still don't like this.  Any time we store the same data in two places, someone new to the codebase updates one while being unaware of the other.  Instead of asserting the two values are the same, let's just store it one.  If, for example, a LeaderSchedule becomes a member variable of frozen banks, this value is redundant.  If, for example, we store `Vec<LeaderSchedule>` from the genesis block, then `current_epoch` is just the array index - again redundant data.  Furthermore, without that field, you could generate a `LeaderSchedule` from just an active set (which requires only a bank) and that `ticks_per_slot` (which may be a constant).

Reply:
The LeaderSchedule is generated by the LeaderScheduler through the generate_schedule() funciton, which requires you pass a bank and a tick_height to seed the random generation of the schedule, so I don't think we can remove that tick_height argument and just leave the bank. The current_epoch is a simple derivation from that tick_height argument. Also without a reference to the starting point of the schedule, it seems messy to force the user of the LeaderSchedule to track how to index into the schedule. Tagging the slot_leaders with Vec<(SlotHeight, LeaderSchedule)> seems clumsier than just storing a single u64 in the struct

Reply:
@carllin 
Given a sorted active set, the node at any given slot is easy to compute.

generate a vector of all the weight sums up to that node.
generate a random number that is an Fnv hash of the epoch and the slot index % last weight
scan the weight vector until the next node is greater then the random number

i was playing around with this
https://gist.github.com/aeyakovenko/f10aa07ce00905fabaeb9ce832c97138

Reply:
That's the separation I was going for too.  For the algorithm you described, I'd put that here:

https://github.com/solana-labs/solana/pull/2841/files#diff-ace2e1305f293d2eff4ec4e9ab859271R55

Reply:
@aeyakovenko, so are we ditching the round robin style of leader scheduling and instead going with a more stake weighted approach?

Reply:
Kind of the same thing if the seed is known in advance.